executor/cook/executor.py

```diff
@@ -0,0 +1,350 @@
+    """Awaits process (available in process_info) completion.
+
+    Parameters
+    ----------
+    driver: MesosExecutorDriver
+        The mesos driver to use.
+    task: map
+        The task that is executing.
+    stop_signal: Event
+        Event that determines if an interrupt was sent
+    process_info: tuple
+        Tuple of (process, stdout_file, stderr_file)
+    shutdown_grace_period_ms: int
+        Grace period before forceful kill
+
+    Returns
+    -------
+    True if the process was killed, False if it terminated naturally.
+    """
+    process, _, _ = process_info
+
+    process_killed = False
+    while is_running(process):
+
+        if stop_signal.isSet():
+            logging.info('Executor has been instructed to terminate')
+            kill_task(driver, task, process_info, shutdown_grace_period_ms)
+            process_killed = True
+            break
+
+        time.sleep(1)
+    return process_killed
+
+
+def manage_task(driver, task, stop_signal, completed_signal, config, stdout_name, stderr_name):
+    """Manages the execution of a task waiting for it to terminate normally or be killed.
+       It also sends the task status updates, sandbox location and exit code back to the scheduler.
+       Progress updates are tracked on a separate thread and are also sent to the scheduler.
+       Setting the stop_signal will trigger termination of the task and associated cleanup.
+
+    Returns
+    -------
+    Nothing
+    """
+    task_id = get_task_id(task)
+    try:
+        sandbox_message = json.dumps({'sandbox-location': config.sandbox_location, 'task-id': task_id})
+        send_message(driver, sandbox_message, config.max_message_length)
+
+        process_info = launch_task(driver, task, stdout_name, stderr_name)
+        process, _, _ = process_info
+
+        progress_watcher = cp.ProgressWatcher(config, completed_signal)
+        progress_updater = cp.ProgressUpdater(config.progress_sample_interval_ms, send_message)
+        progress_complete_event = cp.launch_progress_tracker(driver, task_id, config.max_message_length,
+                                                             progress_watcher, progress_updater)
+
+        process_killed = await_process_completion(driver, task, stop_signal, process_info,
+                                                  config.shutdown_grace_period_ms)
+        exit_code = process.returncode
+
+        # set the completed signal to trigger progress updater termination and await progress
+        # updater termination if executor is terminating normally
+        completed_signal.set()
+        if not stop_signal.isSet():
+            logging.info('Awaiting progress updater completion with timeout of one second')
+            progress_complete_event.wait(1)
+
+        logging.info('Process completed with exit code {}'.format(exit_code))
+        exit_message = json.dumps({'exit-code': abs(exit_code), 'task-id': task_id})
```

Comment:
By using `abs` you may collide with an exit code that has meaning with the program.  Let's just leave them negative.